sklearn/metrics/pairwise.py

```diff
@@ -550,8 +563,249 @@ def cosine_distances(X, Y=None):
+                hi = np.nanmax([hi, np.nanmax(col_array)])
+
+            if np.isnan(lo):
+                lo = 0.0
+            if np.isnan(hi):
+                hi = 0.0
+
+            min_of_numeric[col] = lo
+            max_of_numeric[col] = hi
+
+        # Normalizes the numeric values between 0 and 1.
+        if normalize:
+            scaler = MinMaxScaler().fit([min_of_numeric, max_of_numeric])
+            X_num = scaler.transform(X_num)
+            if not y_none:
+                Y_num = scaler.transform(Y_num)
+            else:
+                Y_num = X_num
+        # If not, checks if the data is normalized.
+        elif np.min(min_of_numeric)<0 or np.max(max_of_numeric)>1:
+            warnings.warn("Input data is not normalized between 0 and 1.", UserWarning)
+
+    y_n_rows = Y.shape[0]
+    n_col_num = X_num.shape[1]
+    n_col_cat_obj = X_cat_obj.shape[1]
+    n_col_cat_num = X_cat_num.shape[1]
+
+    D = np.zeros((n_rows, y_n_rows), dtype=np.float32)
+
+    for i in range(n_rows):
+        j_start = i
+
+        # for non square results
+        if n_rows != y_n_rows or not y_none:
+            j_start = 0
+
+        sum_s = 0
+        n_cols_not_missing = np.zeros(n_rows - j_start)
+        # sum and missing values treatment for each type of data, if available
+        if n_col_cat_obj > 0:
+            row_cat_none = ~(np.equal( X_cat_obj[i, :], None) |
+                             np.equal(Y_cat_obj[j_start:n_rows, :], None))
+
+            n_cols_not_missing = np.sum(row_cat_none.astype(np.int), axis=1)
+            sum_s = (~np.equal( X_cat_obj[i, :], Y_cat_obj[j_start:n_rows, :])
+                     & row_cat_none).astype(np.int).sum(axis=1)
+
+        if n_col_cat_num > 0:
+            row_cat_nan = ~(np.isnan(X_cat_num[i, :]) |
+                            np.isnan(Y_cat_num[j_start:n_rows, :]))
+
+            n_cols_not_missing = n_cols_not_missing + \
+                                 np.sum(row_cat_nan.astype(np.int), axis=1)
+
+            sum_s = sum_s + (~np.equal(X_cat_num[i, :],
+                                       Y_cat_num[j_start:n_rows, :])
+                             & row_cat_nan).astype(np.int).sum(axis=1)
+
+        if n_col_num > 0:
+            n_cols_not_missing = n_cols_not_missing + \
+                                 np.sum((~(np.isnan(X_num[i, :]) |
+                                           np.isnan(Y_num[j_start:n_rows, :])))
+                                        .astype(np.int), axis=1)
+
+            sum_s = sum_s + np.nansum(
+                np.absolute(X_num[i, :] - Y_num[j_start:n_rows, :]), axis=1)
+
+        result = np.divide(sum_s, n_cols_not_missing,
+                           out=np.full(n_cols_not_missing.shape, np.nan),
+                           where=n_cols_not_missing != 0)
+
+        D[i, j_start:] = result
+        if y_none:
+            D[i:, j_start] = result
+
+    return D
+
+def detect_categorical_features(X, categorical_features=None):
```

Comment:
I'm not sure we want the magic provided by this function. This seems designed to work with arrays of dtype object. It is not logic we have elsewhere, and for now I would like to support numeric arrays where categorical features are specified. If we do want this logic, we should be making it a helper in sklearn.utils to generally deal with categorical feature determination in object arrays and DataFrames.

Reply:
Should I remove this function and make the categorical_features parameter mandatory, push it back inside the gower function, or move it to sklearn.utils with support to DataFrames?

Reply:
For now make categorical features mandatory, but we should make it work just like the column specification in ColumnTransformer, allowing it to be a function.

Reply:
I just realized this function is too complex and coupled with Gower function, to push to a user to develop. We need it to identify categorical columns represented as object, categorical values represented as integers, then I define the numerical columns